### PR TITLE
feat: Abstract Frobenius elements

### PR DESCRIPTION
**Pull request by kbuzzard**
This PR will be closed because it's not in the right generality; the current approach is outlined [here](https://leanprover.zulipchat.com/#narrow/stream/416277-FLT/topic/Outstanding.20Tasks.2C.20V4/near/449562398). 

We develop the "abstract" theory Frobenius elements. The set-up is a commutative ring A, a commutative A-algebra B, and maximal ideals P of A under Q of B. Assuming that the fixed points of Aut_{A-alg}(B) are A, and that B/Q and A/P are finite, we prove the existence of a Frobenius element in Aut_{A-alg}(B). This is the "guts" of the construction; applications to number fields and function fields will come later.

---

- [x] depends on: #13294

[Open in Gitpod](https://gitpod.io/from-referrer/)

Sorry that this PR is long, although note that the first 100 lines are mostly docstrings.

### CODE REVIEW

File: Mathlib/NumberTheory/Frobenius/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+  rw [smul_eq_iff_eq_inv_smul]
+  by_contra h
+  have hy : y A Q ∈ (Frob Q isGalois P)⁻¹ • Q := (y_spec A Q).2 _ ⟨_, rfl⟩ (Ne.symm h)
+  have hy2 : (Frob Q isGalois P) • (y A Q) ∈ Q := by
+    rwa [Ideal.pointwise_smul_eq_comap] at hy
+  have this := Q.sub_mem hy2 <| Frob_spec Q isGalois P
+  simp only [sub_sub_cancel] at this
+  apply y_not_in_Q A Q <| Ideal.IsPrime.mem_of_pow_mem (show Q.IsPrime by infer_instance) _ this
+
+lemma Frob_Q_eq_pow_card (x : B) : Frob Q isGalois P x - x ^ (Fintype.card (A⧸P)) ∈ Q := by
```
Comment (kbuzzard) [inline Mathlib/NumberTheory/Frobenius/Basic.lean:287]:
```suggestion
lemma Frob_eq_pow_card (x : B) : Frob Q isGalois P x - x ^ (Fintype.card (A⧸P)) ∈ Q := by
```

File: Mathlib/NumberTheory/Frobenius/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+lemma g_spec : ∀ (z : (B ⧸ Q)ˣ), z ∈ Submonoid.powers (g Q) :=
+  (IsCyclic.exists_monoid_generator (α := (B ⧸ Q)ˣ)).choose_spec
+
+variable [Fintype (B ≃ₐ[A] B)] [DecidableEq (Ideal B)]
+
+/-- An element `y` of `B` exists, which is congruent to `b` mod `Q`
+and to `0` mod all other Galois conjugates of `Q` (if any).-/
+lemma exists_y :
+    ∃ y : B, (y : B ⧸ Q) = g Q ∧ ∀ Q' :
+    Ideal B, Q' ∈ MulAction.orbit (B ≃ₐ[A] B) Q → Q' ≠ Q → y ∈ Q' := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/Frobenius/Basic.lean:124]: It looks odd to split the line in the middle of `∀ Q' : Ideal B`

File: Mathlib/NumberTheory/Frobenius/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+    convert Ideal.comap_isMaximal_of_surjective (K := Q) _ (AlgEquiv.surjective σ.symm)
+    apply Ideal.pointwise_smul_eq_comap
+  have hPairwise : Pairwise fun (I : O) (J : O) ↦ IsCoprime (I : Ideal B) J := fun x y h ↦ ⟨1, 1, by
+    simp only [Ideal.one_eq_top, Ideal.top_mul]
+    exact Ideal.IsMaximal.coprime_of_ne (hmax x) (hmax y) <| mt Subtype.ext h⟩
+  obtain ⟨y, hy⟩ := Ideal.exists_forall_sub_mem_ideal (ι := O) hPairwise
+    (fun J ↦ if J = ⟨Q, 1, by simp⟩ then (Ideal.Quotient.mk_surjective (g Q : B ⧸ Q)).choose else 0)
+  refine ⟨y, ?_, ?_⟩
+  · specialize hy ⟨Q, 1, by simp⟩
+    simp at hy
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/Frobenius/Basic.lean:138]: Please squeeze

File: Mathlib/NumberTheory/Frobenius/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+/-- `F : B[X]` is defined to be a product of linear factors `(X - τ • y)`; where
+`τ` runs over `B ≃ₐ[A] B`, and `y : B` is an element which generates `(B ⧸ Q)ˣ`
+and lies in `τ • Q` for all `τ • Q ≠ Q`.-/
+noncomputable abbrev F : B[X] := ∏ τ : B ≃ₐ[A] B, (X - C (τ • (y A Q)))
+
+lemma F_spec : F A Q = ∏ τ : B ≃ₐ[A] B, (X - C (τ • (y A Q))) := rfl
+
+variable {A Q} in
+open Finset in
+lemma F.smul_eq_self (σ :  B ≃ₐ[A] B)  : σ • (F A Q) = F A Q := calc
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/Frobenius/Basic.lean:179]:
```suggestion
lemma F.smul_eq_self (σ : B ≃ₐ[A] B) : σ • (F A Q) = F A Q := calc
```

File: Mathlib/NumberTheory/Frobenius/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+  rw [← aeval_map_algebraMap B, m_spec', algebraMap.coe_def, aeval_algebraMap_apply,
+    coe_aeval_eq_eval, F.y_eq_zero A Q, map_zero]
+
+open scoped Polynomial
+
+variable [Fintype (A⧸P)]
+
+lemma m.mod_P_y_pow_q_eq_zero' :
+    aeval ((algebraMap B (B⧸Q) (y A Q)) ^ (Fintype.card (A⧸P)))  (m A Q isGalois) = 0 := by
+  let foobar : Field (A⧸P) := ((Ideal.Quotient.maximal_ideal_iff_isField_quotient P).mp ‹_›).toField
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/Frobenius/Basic.lean:249]: Can we find a better name for this?

File: Mathlib/NumberTheory/Frobenius/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+  rw [← aeval_map_algebraMap (A⧸P), FiniteField.aeval_pow_card, ← algebraMap.coe_def,
+    aeval_map_algebraMap, m.y_mod_P_eq_zero, zero_pow Fintype.card_ne_zero]
+
+lemma F.mod_Q_y_pow_q_eq_zero' :
+    aeval ((algebraMap B (B⧸Q) (y A Q)) ^ (Fintype.card (A⧸P))) (F A Q) = 0 := by
+  rw [← m_spec' A Q isGalois, aeval_map_algebraMap, m.mod_P_y_pow_q_eq_zero']
+
+lemma _root_.Polynomial.aeval_finset_prod.{u, v, y} {R : Type u} {S : Type v} {ι : Type y}
+    [CommSemiring R] [CommSemiring S] [Algebra R S] (s : Finset ι) (g : ι → R[X]) (x : S) :
+  aeval x (∏ i ∈ s, g i) = (∏ i ∈ s, aeval x (g i)) := eval₂_finset_prod (algebraMap R S) s g x
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/Frobenius/Basic.lean:259]:
```suggestion
    aeval x (∏ i ∈ s, g i) = (∏ i ∈ s, aeval x (g i)) := eval₂_finset_prod (algebraMap R S) s g x
```

Comment (kbuzzard) [reply Mathlib/NumberTheory/Frobenius/Basic.lean:259]: I have mistakenly left this lemma here -- probably it should go elsewhere.

File: Mathlib/NumberTheory/Frobenius/Basic.lean [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+variable (P : Ideal A) [P.IsMaximal] [Algebra (A ⧸ P) (B ⧸ Q)] [IsScalarTower A (A⧸P) (B⧸Q)]
+variable {A} -- now can be implicit because `P`
+
+lemma m.y_mod_P_eq_zero : Polynomial.aeval (↑(y A Q) : B ⧸ Q) (m A Q isGalois) = 0 := by
+  rw [← aeval_map_algebraMap B, m_spec', algebraMap.coe_def, aeval_algebraMap_apply,
+    coe_aeval_eq_eval, F.y_eq_zero A Q, map_zero]
+
+open scoped Polynomial
+
+variable [Fintype (A⧸P)]
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/Frobenius/Basic.lean:245]: Same here.

File: Mathlib/NumberTheory/Frobenius/Basic.lean [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+  rw [F_spec, aeval_finset_prod, Finset.prod_eq_zero_iff] at this
+  obtain ⟨σ, -, hσ⟩ := this
+  use σ
+  simp only [Ideal.Quotient.algebraMap_eq, AlgEquiv.smul_def, map_sub, aeval_X, aeval_C,
+    sub_eq_zero] at hσ
+  exact (Submodule.Quotient.eq Q).mp (hσ.symm)
+
+/-- An auxiliary arithmetic Frobenius element, in the automorphism group of the integer ring
+rather than the global field itself. -/
+noncomputable abbrev Frob := (exists_Frob Q isGalois P).choose
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/Frobenius/Basic.lean:268]: The full name is `ArithmeticFrobenius.Abstract.Frob`. I understand you want to make it clear we are talking about the arithmetic one, but `Abstract` seems useless.

Comment (kbuzzard) [reply Mathlib/NumberTheory/Frobenius/Basic.lean:268]: But we will never use this Frob, as it is an element of Aut(B/A), not Gal(L/K), and one of its inputs is some weird hypothesis about everything in B which is Aut-stable is in A. *Everything* in the Abstract namespace is useless! Actual Frobenius elements are the next PR. As far as I am concerned, this definition could be private!

Comment (riccardobrasca) [reply Mathlib/NumberTheory/Frobenius/Basic.lean:268]: I see, no problem then!.

Comment (kbuzzard) [reply Mathlib/NumberTheory/Frobenius/Basic.lean:268]: How about I call the namespace Aux?

Comment (riccardobrasca) [reply Mathlib/NumberTheory/Frobenius/Basic.lean:268]: This seems a good idea!

**Comment by kbuzzard** [comment]
Joel Riou has pointed out [here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Uses.20of.20Frobenius.20elements.20in.20mathematics/near/448934112) that this argument generalizes to prove a cleaner statement (no finiteness assumptions) so I will modify this PR to prove the even more general thing (apparently it's useful for the etale fundamental group; the argument is very similar to what is here, but uses integrality more).

**Comment by kbuzzard** [comment]
Closed in favour of #17717

**Pull request closed without merging**